schema/union.go

```diff
@@ -7,22 +7,38 @@ import (
 )
 
 type UnionField struct {
-	name       string
-	itemType   []AvroType
-	definition []interface{}
+	name           string
+	avroTypes      []AvroType
+	generatedTypes []AvroType
+	definition     []interface{}
+	optTypeIdx     int
 }
 
-func NewUnionField(name string, itemType []AvroType, definition []interface{}) *UnionField {
-	return &UnionField{
+func NewUnionField(name string, avroTypes []AvroType, definition []interface{}, optTypeIndex int) *UnionField {
+	u := &UnionField{
 		name:       name,
-		itemType:   itemType,
+		avroTypes:  avroTypes,
```

Comment:
Some of these changes (like renaming existing variables) could be split into a seperate PR to make this one smaller. It makes it tougher to know what's related to the nullable fields change and what's refactoring existing code.

Reply:
I agree.